rfd/0030-upgrade-system.md

```diff
@@ -0,0 +1,444 @@
+but we don't want to orphan older versions, so lets err on the side of caution here.
+
+Outstanding design questions:
+- Grpc or basic https API?  My preference leans toward grpc, but its hard to discount the possible
+benefits of a basic https API (e.g. pulling down release info with `curl`).
+- How 'smart' should the index server be exactly?  On one hand, if the index server is able to intelligently
+select what releases/targets are relevant to a given client, it may be able to reduce the impact of bugs
+in the upgrade system (easier to update the index server than to update all clients).  On the other hand,
+the extensibility of the upgrade system (not to mention the simplicity of the index server) would likely
+be benifited by keeping as much logic as possible in the client clusters.
+
+Hypothetical teleport release descriptors returned by index server:
+
+```yaml
+targets:
+  - tags: [oss]
+    version: '100.1.0-alpha.2'
+    arch: amd64
+    os: linux
+    flavor: oss
+    stable: false
+    security_patch: false
+    deprecated: false
+    sums:
+      rpm-blake2-256: '...'
+      tgz-blake2-256: '...'
+      deb-blake2-256: '...'
+  - tags: [ent]
+    version: '100.0.1'
+    arch: amd64
+    flavor: ent
+    os: linux
+    stable: true
+    security_patch: true
+    deprecated: false
+    sums:
+      rpm-blake2-256: '...'
+      tgz-blake2-256: '...'
+      deb-blake2-256: '...'
+  - tags: [ent,fips]
+    version: '100.0.1'
+    arch: amd64
+    os: linux
+    flavor: ent
+    stable: true
+    security_patch: true
+    deprecated: false
+    sums:
+      rpm-blake2-256: '...'
+      tgz-blake2-256: '...'
+      deb-blake2-256: '...'
+```
+
+*note*: Since writing the above I'm leaning toward grouping targets by version number with a single set of user-assignable
+tags/labels stored at the top-level of the version entry.  Will circle back and update.
+
+#### Upgrade Policies
+
+An upgrade policy will be responsible for matching releases to servers.  In the interest of
+reliability and fault-taulerance, many upgrade policies may apply to a given server.  The
+upgrade policy which maps to the most recent teleport version will be given priority for
+a given server.
+
+Policy evaluation (i.e. determining how releases map to servers) will be the responsibility
+of the auth server.  The goal of this centralization is to ensure that if a logic bug in the
+mapping results in requiring a manual upgrade, only the auth server(s) will need to be upgraded.
+Some amount of hedging will be added to the API to make sure its easy to add custom (external)
+policy evaluators later if the need exists, but I imagine that custom upgrade controllers and
+manually managed release manifests will cover most usecases.
+
+I have yet to work out the exact details of how policy syntax and evaluation should work, but
+here's an example of one of the syntaxes I've been toying with in my head:
+
+```yaml
+target_selectors:
```

Comment:
I like the policy approach. What happens if I update the policy on the fly during upgrade? Is there a transactional state to updating policies?

Reply:
Added upgrade UX section which covers my current thinking on this.